### PR TITLE
[Refactor] XContent base classes from xcontent to core library

### PR DESCRIPTION
**Pull request by nknize**
Refactors the xcontent base classes from the xcontent module to the core library so serializable xcontent contracts can be used across other libraries (e.g., new aggregation library, other community contributed libraries). Since `opensearch-core` is intended to be the foundation library for all of opensearch, this PR continues the effort to refactor those foundation classes from the `:server` module to appropriate library and base module plugins in order to reduce the surface area of the `:server` module.

Note: The surface area of this change is large due to xcontent being used as a foundation across the entire project. 

relates #5910 
depends on #6218

### CODE REVIEW

**Comment by andrross** [comment]
Am I correct in assuming that the Java package change (from `org.opensearch.common` to `org.opensearch.core`) makes this a breaking change that cannot be backported to the 2.x branch?

**Comment by reta** [comment]
@nknize I have difficulties understanding the reason for this change

> Refactors the xcontent base classes from the xcontent module to the core library so serializable xcontent contracts can be used across other libraries

What exactly is stopping other libraries to use `x-content` module? We publish it as an artifact for everyone to use, I think consolidating all related `x-content` classes (pulling from core & co if needed) would be the way to shape this change.

>  Since opensearch-core is intended to be the foundation library for all of opensearch, this PR continues the effort to refactor those foundation classes from the :server module to appropriate library and base module plugins in order to reduce the surface area of the :server module.

We have more or less concise set of modules, why do we want to beef everything into `opensearch-core` vs keeping a set of modules? Also, there is more profound issue that this change introduces (or better to say, amplifies) the packages split brain: the `o.o.c.xcontext` package is now split between different modules, a showstopper for any future modularity efforts.

**Comment by nknize** [comment]
> ....a showstopper for any future modularity efforts.

On the surface it seems that way so let me address each of these questions separately.

> What exactly is stopping other libraries to use x-content module?

The build is [configured](https://github.com/opensearch-project/OpenSearch/blob/main/libs/build.gradle#L39) such that library modules may not depend on other library modules except `opensearch-core`.  This heavily restricts us from using libraries to trim down the bloated `:server` module (I'll give an example below). It was designed this way so that core retains generics and abstract contracts used across all of opensearch with decoupled concrete implementations in specific libraries, modules, and plugins.

> We have more or less concise set of modules, why do we want to beef everything into opensearch-core vs keeping a set of modules? 

This change isn't beefing everything into core. It's teasing out the reusable generics and interface contracts to the core library which is used to implement the concrete classes and architecture details across modules and concrete libraries. 

> Also, there is more profound issue that this change introduces (or better to say, amplifies) the packages split brain: the o.o.c.xcontext package is now split between different modules

It seems that way, but if you look at the change it's sticking w/ the intended design and refactoring the XContent base classes and interfaces to the core library while leaving the concrete implementations of those MediaTypes (e.g., `JSON`, `CBOR`, `SMILE`, `YAML`) in the x-content library. As it turns out, we might need to refactor those concrete implementations to an `x-content` module in a follow up PR but I'm deferring that unless it's absolutely necessary.

The driver for this is to further modularize the `:server` module by teasing out the generics to libraries and implementations to modules. Let me give an example where the cross library dependency limitation currently restricts a big part of this (hence the xcontent change).

###  [Aggregations](https://github.com/opensearch-project/OpenSearch/tree/main/server/src/main/java/org/opensearch/search/aggregations) 

Third party libraries, plugins, modules, etc, want to build new aggregations all the time. We provide this ability and it's super powerful. To do so is quite boiler plate: 1. If the [CoreValuesSourceType](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/aggregations/support/CoreValuesSourceType.java) library doesn't already provide the requisite [ValueSource](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/ingest/ValueSource.java), write the needed [ValuesSourceType](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/aggregations/support/ValuesSourceType.java) the new aggregation depends on, 2. write the [Aggregator](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/aggregations/Aggregator.java#L60), 3. write the [Aggregator Factory](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/aggregations/AggregatorFactory.java), 4. write the [AggregationBuilder](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/aggregations/AggregationBuilder.java), 5. [register the aggregation](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/plugins/SearchPlugin.java#L165) through your plugin using the [AggregationSpec](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/plugins/SearchPlugin.java#L436)

As you can see in this process there is a large number of abstract classes and interfaces required to support implementing a new concrete aggregation anywhere in the "ecosystem". Organizationally, these base classes are spread across the [aggregations](https://github.com/opensearch-project/OpenSearch/tree/main/server/src/main/java/org/opensearch/search/aggregations) and [aggregations.support](https://github.com/opensearch-project/OpenSearch/tree/main/server/src/main/java/org/opensearch/search/aggregations/support) along with the mechanisms needed to implement the aggregation phase(s) in the [SearchModule](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/SearchModule.java) and [SearchService](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/SearchService.java). This means for anyone wanting to build a custom aggregation they would take a `build.gradle` dependency on `"org.opensearch:opensearch:${opensearchVersion}"` which pulls in a [bloated server jar](https://mvnrepository.com/artifact/org.opensearch/opensearch/2.4.1) containing much more than the classes and dependencies they need to build their aggregation. 

I think most on this project agree this is a modularization anti-pattern, so to remedy we're striving to make the codebase much more modular. So how do we achieve that here with this clumsy aggregation foundation and user facing "API" (which isn't exactly well documented, or implemented, or supported)? At first pass you might say, "hey, how about we stick with the intent of the libraries and modules design and refactor the generics and base classes from the bloated `:server` module to a new `opensearch-aggregations` library with the concrete implementation in a new `aggs-common` module that continue to use the `search.aggregations.*` namespace? I'd agree with this, in fact I'd go a step further and say this is where I think we should go with, not just aggregations, most of the big opensearch mechanisms (field mappers, field types, doc value types, field data, etc). In the end the `:server` module should be nothing more than the "motherboard" providing the concrete implementation of all the opensearch base components within the libraries and concrete extendable mechanisms provided by the modules and optional plugins. 

But if you embark down that journey you will quickly find the cross library dependency restriction blocks much of this objective and the tight dependency requirement of the `:server` module with other modules quickly leads to JarHell issues. Take the aggregation decoupling use case, for example. Aggregations, like field mappers, are serializable and therefore [depend on XContent](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/aggregations/Aggregation.java#L45) in order to implement the ToXContentFragment logic. So how about refactoring all the XContent base classes to the `opensearch-x-content` library then? That works until you refactor any base interfaces to a library other than `opensearch-core`, because libraries cannot depend on each other. 

So that leaves us with a few options:

1. remove the cross library dependency restriction
2. refactor all base classes that would result in a cross library dependency situation into modules instead
3. Do nothing, leave all base classes and interfaces that would result in a cross library dependency in the server module and leave users to depend on the massive jar
4. refactor opensearch core base classes and interface contracts that would result in a cross library dependency to the `opensearch-core` library and the implementation details to other libraries, modules, and plugins

With this PR I chose Option 4 because it not only fits the modularization objective but after the aggregation framework is refactored to a new module (and supporting library) users in the aggregation scenario above would only need to take a dependency on the opensearch-core and a (soon coming) opensearch-aggregations library to implement a new concrete aggregation in their own plugin.

I hope this helps paint a clearer picture?

**Comment by reta** [comment]
> I hope this helps paint a clearer picture?

Thanks @nknize , yes, it does. I missed the restriction on `opensearch-core` usage. 

> 4. refactor opensearch core base classes and interface contracts that would result in a cross library dependency to the opensearch-core library and the implementation details to other libraries, modules, and plugins

Now it make sense, I think we should not open the gate for cross library dependency restriction (I believe you also don't want to do that) but what about something in the middle: introduce `opensearch-core-xcontent` module and have a set of core modules (not just `opensearch-core`), the `opensearch-core-aggregations` could follow fe? 

I am afraid that the limitation on just one `opensearch-core` would lead to it becoming a kitchen sink of everything the plugin / extension / ... may need.

**Comment by andrross** [comment]
> > Am I correct in assuming that the Java package change (from org.opensearch.common to org.opensearch.core) makes this a breaking change that cannot be backported to the 2.x branch?
> 
> This change intends to target 3.0 only (I just added the labels, sorry about that). However, that wouldn't preclude a backport if we decided to pull it into 2.x since mixed cluster rolling upgrades are only mixed nodes not mixed jars. The side effect, however, is a namespace ripple across all plugins that depend on xcontent base classes. I'd wager that's at least 90% of plugins. The refactor would have to be done before releasing the 2.x version that would include this change.

I think the namespace change would preclude a backport since it would violate our compatibility guarantees even if the change is trivial for plugins. Can this change be done without the Java package name change in a way that makes it backward compatible?

**Comment by nknize** [comment]
> introduce opensearch-core-xcontent module and have a set of core modules

This would work as a intermediate step. However, my hope was something along the following (let me know your thoughts):

1. refactor `CoreValuesSourceType` and ValuesSource implementations to `opensearch-core` 
2. a new `:aggregation-framework` module that implements the core aggregation framework (services, plumbing, etc).
3. a new `:aggregations-common` module that implements the core aggregations (e.g., `avg`, `sum`, `histogram`, `extended_stats`, basically all the aggs in the `aggregations` namespace)
4. a new `:opensearch-aggregations` library that provides the base classes, interfaces, contracts used by 2. and 3. above

With the above design the two new modules take a dependency on the `opensearch-core` and `opensearch-aggregations` libraries to implement the framework and core aggregations provided by opensearch min distribution. Users then only need take a dependency on the `opensearch-core` and `opensearch-aggregations` libraries to implement their concrete implementation. 

By moving everything into modules users would need to take a dependency on the modules which brings in concrete implementations they may not care about.


> I am afraid that the limitation on just one opensearch-core would lead to it becoming a kitchen sink of everything the plugin / extension / ... may need.

I agree this could be a risk, but that's already the case w/ `:server` module today because there is no clear definition of what goes where. I'm hoping this refactor PR trailblazes those decisions so we can document this in the `package-info.java` javadocs and enforce in our PR reviews. To track and facilitate ongoing discussion about future refactor PRs I opened the #5910 meta issue.

**Comment by nknize** [comment]
@andrross 

> violate our compatibility guarantees

Are you referring to a cosmetic guarantee that a dependency namespace doesn't change w/in minor releases? Remind me if we have this documented somewhere?

**Comment by reta** [comment]
> > Are you referring to a cosmetic guarantee that a dependency namespace doesn't change w/in minor releases? Remind me if we have this documented somewhere?
> 
> [This is the most extensive semver documentation](https://opensearch.org/blog/what-is-semver/) I'm aware of. I think it comes down to interpretation of "API". If we consider these XContent classes as part of the API used by plugins, then a namespace change is certainly a breaking change to that API.

I would go further, moving classes to different JARs even under the same package is a breaking change: we could not guarantee that some use only subset of JARs and as such, the classes won't be there anymore, requiring to pull in new dependency.

**Comment by nknize** [comment]
> ...moving classes to different JARs even under the same package is a breaking change

It is. I think the key here, though, is whether it breaks a "forward facing API" (as defined by the project) that formally complies w/ SemVer. We're allowed to "break" `@opensearch.internal` and `@opensearch.experimental` classes because they're not classified as forward facing `@opensearch.api`. I think that's the "one way door" change that needs to be thought about and maybe github labeled in PR reviews. Longer term I'd like opensearch to have something more enforceable and automated like Apache Spark's [Developer Annotations](https://github.com/apache/spark/tree/master/common/tags/src/main/java/org/apache/spark/annotation). I added the javadoc tags as a first step for the missingJavadoc task because there wasn't anything else. Longer term I think we can add more automation and better enforcement through gradle checks using a well defined annotation implementation.

**Comment by reta** [comment]
> With the above design the two new modules only take a dependency on the `opensearch-core` and `opensearch-aggregations` libraries to implement the framework and core aggregations provided by opensearch min distribution.

Aha, the bigger picture starts to emerge, I think it is a viable path ahead. 

> I agree this could be a risk, but that's already the case w/ :server module today because there is no clear definition of what goes where. 

That we need to clean up I think.

Thanks @nknize , could I ask you please to formalize your decisions in the issue https://github.com/opensearch-project/OpenSearch/issues/5910 (and create sub-issues if necessary), that could help other to understand the path forward. The only note I would like to have (this is a good time) is to ensure that the core modules (opensearch-core) and the implementation modules (x-content) do not have package  split-brains. I won't matter much right now but would eliminate yet another set of breaking changes in the future. 

Thank you.

**Comment by nknize** [comment]
> ...could I ask you please to formalize your decisions in the issue https://github.com/opensearch-project/OpenSearch/issues/5910 (and create sub-issues if necessary)

Definitely! I'll copy the details to the meta issue and reference this conversation for posterity. 

> The only note I would like to have (this is a good time) is to ensure that the core modules (opensearch-core) and the implementation modules (x-content) do not have package split-brains.

I agree. This should be formally written down somewhere (maybe CONTRIBUTING.MD?) and somehow enforced in the PR reviews. That will be difficult and we'll likely end up w/ future PRs introducing a split brain in the meantime. At least until all maintainers (and future maintainers) develop the muscle memory to check for this on PR review.

**Comment by andrross** [comment]
I'll go ahead and explicitly state the concern I was getting at with my questions around backporting: given that the [3.0 release is not yet scheduled](https://opensearch.org/releases.html), I don't love the idea of a structural change that impacts so many files like this. We've already had problems with divergence between the `main` and `2.x` branches causing complications in development (/cc @kartg) and this seems likely to exacerbate that problem. I'd prefer to see changes like this deferred until there is at least a timeframe for 3.0, especially if 3.0 may not be released for over another year from now.

**Comment by nknize** [comment]
> I don't love the idea of a structural change that impacts so many files like this. We've already had problems with divergence between the `main` and `2.x` branches causing complications in development.

Let's work through that here because we shouldn't shy away from large progress (especially for so long) because backporting is "complicated". We should figure out inventive ways to reduce the friction to enable incremental progress. Even if that means we figure out better ways to use worktrees or some other guy mechanism.

**Comment by andrross** [comment]
Fundamentally our branching strategy somewhat favors development on the next major version at the cost of backporting to a divergent branch for the current major version, and I think that's the thing I'm poking at here. I'm definitely not an expert in this area and changing that strategy is a bigger conversation than this specific change. I admit holding off on desirable changes to avoid divergence is really bad as it impedes progress. Though I think the technical solutions here are likely limited because at some point human judgment will be needed to apply a logical change across divergent code bases. Maybe other folks don't see it as a big problem, but I do worry about introducing bugs anytime a developer has to do non-trivial conflict resolutions across branches.

**Comment by reta** [comment]
> Maybe other folks don't see it as a big problem, but I do worry about introducing bugs anytime a developer has to do non-trivial conflict resolutions across branches.

I share your concern @andrross , the change of such a scope would basically make backporting automation useless (except for trivial cases) - it will always end up either with conflicts or compilation error.

From the other side, we have to move forward, so what option do we have? 
1. Move the bits between modules but no package changes: significantly smaller impact area
2. It would help now but only delay the package change for later 

Or we could try to:
1. Move the bits to the `core` package
2. But keep the old classes around (same package but deprecate for removal), they would be proxies that delegate to the `core` ones 
3. In 3.0.0 we could remove the old classes

Just some thoughts ....

**Comment by CEHENKLE** [comment]
I really like the direction this is going in, but the timing is giving me some heartburn.  What if we held off on merging this change until we had a date for 3.0 (~6-8 weeks out).  This would minimize the amount of time we have where developers would have to straddle (for want of a better word) this change with the risk of making a mistake. 

In a related question, do you have any reading of the tealeaves, Nick, when the next version of Lucene might be coming out?

**Comment by nknize** [comment]
> In a related question, do you have any reading of the tealeaves, Nick, when the next version of Lucene might be coming out?

My gut says not for a while. 9.5 release was just proposed a week ago w/ majority favoring. History is trending toward a longer tail between major Lucene releases. Which could be a good thing for OpenSearch decoupling major releases from Lucene changes. Also gives reasonable time to think through Migrations and BWC such that we could absorb lucene index breaking changes across minor releases.

**Comment by nknize** [comment]
> > From the other side, we have to move forward, so what option do we have?
> > 
> > 1. Move the bits between modules but no package changes: significantly smaller impact area
> > 2. It would help now but only delay the package change for later
> 
> @reta This is basically the approach we've taken with the inclusive naming project. There are some very widely used classes (e.g. `org.opensearch.action.support.master.AcknowledgedResponse`) that for various reasons can't really live side-by-side with a renamed variant, so we've deprecated that package name but are waiting to do the very intrusive s/master/clustermanager refactoring until we have a better idea of when the last minor release of the 2.x branch will be.

I lean this direction as well. I'd be fine refactoring this PR to retain the namespace and defer that to a later PR. This change is mostly around moving the extensibility needle and reducing the surface area of `:server` while fostering a thoughtful discussion on reduced impact ways of doing that. It seems to be trending that direction so I thank you all for the thoughtful comments.

**Comment by andrross** [comment]
> I'd be fine refactoring this PR to retain the namespace and defer that to a later PR.

If we take this approach, then we're moving some classes from `:server` to `:core`. Given that any project currently getting these dependencies from `:server` will still get them (transitively) from `:core` after the refactoring, I will argue that this is in fact backwards compatible. @reta Do you agree or am I playing too fast and loose with compatibility concerns?

**Comment by reta** [comment]
> @reta Do you agree or am I playing too fast and loose with compatibility concerns?

@andrross In this particular case I agree with your conclusions: the `:server` has to come with `:core` as the dependency (and `:x-content` also comes with `:core` as well), so we should not break anyone (assuming no package changes).

**Comment by nknize** [comment]
One thing I forgot to mention is that if/when we commit to using the jigsaw modularity framework we have to ensure unique package names within modules. The jigsaw framework does not allow packages to be split between modules. If we define the namespace refactor as a breaking change then I highly suggest we do this before 3.0, so this decision will only be temporary while we tidy up our house by moving classes to appropriate libraries.

**Comment by nknize** [comment]
opened #6218 which will reduce the surface area change of this PR

File: client/rest-high-level/src/test/java/org/opensearch/client/IngestClientIT.java
Comment (reta) [inline client/rest-high-level/src/test/java/org/opensearch/client/IngestClientIT.java:82]: The type cast is not needed anymore, right? (MediaType is just fine)

Comment (nknize) [reply client/rest-high-level/src/test/java/org/opensearch/client/IngestClientIT.java]: Correct. There were a few of those. I pushed a commit to remove all of those superfluous casts.

File: modules/lang-mustache/src/main/java/org/opensearch/script/mustache/SearchTemplateResponse.java
Comment (reta) [inline modules/lang-mustache/src/main/java/org/opensearch/script/mustache/SearchTemplateResponse.java:109]: `MediaType contentType = parser.contentType();` ?

File: modules/reindex/src/test/java/org/opensearch/index/reindex/RestReindexActionTests.java
Comment (reta) [inline modules/reindex/src/test/java/org/opensearch/index/reindex/RestReindexActionTests.java:77]: It seems like `withContent` could be changed to accept `MediaType`, it does not need `XContentType` at all

File: modules/lang-painless/src/main/java/org/opensearch/painless/action/PainlessExecuteAction.java
Comment (reta) [inline modules/lang-painless/src/main/java/org/opensearch/painless/action/PainlessExecuteAction.java:185]: These typecases could bite as in the future, I am wondering should we add the method to `XContentType`:

```
    public static XContentType fromMediaType(MediaType mediaType) {
        if (mediaType instanceof XContentType) {
            return (XContentType) mediaType;
        } else {
            return MEDIA_TYPE_PARSER.fromMediaType(mediaType.mediaTypeWithoutParameters());
        }
    }
```

In this case all typecast will be isolated.

Comment (nknize) [reply modules/lang-painless/src/main/java/org/opensearch/painless/action/PainlessExecuteAction.java]: Yes! I like that idea. Done.

**Review by reta — APPROVED** [review]

**Comment by nknize** [comment]
Gradle check failure is unrelated and looks like it could be a bug in jackson-dataformat-yaml:2.14.2 or snakeyaml:1.32.  Opened bug report in #6411. Refiring check and will address separately.

**Comment by AndreVirtimo** [comment]
This is a breaking change in version 2.10!

Where is the communication of breaking changes? This should be part of the changelog.

**Comment by reta** [comment]
> This is a breaking change in version 2.10!

@AndreVirtimo you are absolutely right, the change was slipped away but it should have been in the change log and release notes for 2.10, sincere apologies for that.

**Comment by nknize** [comment]
> This is a breaking change in version 2.10!

Can you share your use case? Where / how did you run into this? Did you get a NoClassDefFound? Do you have an implementation that relies on it? Are you importing dependencies yourself? 

The reason I ask is because the core technically does not (yet) provide a formal plugin API (we're working on this; feel free to track history through #5910) and dependency libraries are (should) already be handled by the min-distribution, bundle, or docker releases. I want to make sure you didn't stumble on a release artifact bug.

**Comment by olfuerniss** [comment]
@nknize A breaking change info for developers would be important (+ necessary migration steps). 

E.g. for a OpenSearch plugin, you must adjust the plugin version for the target OpenSearch version. This means also you must re-compile the plugin. This will break due to this and the other related refactoring. In our plugin, nearly every class (> 200) had to be adjusted. Not only moved classes but also API changes happened in 2.10. (Strings.toString(XContentBuilder) removed etc.)

But also in other code that uses the REST High Level Client we use the XContentBuilder quite often and like it. The RestStatus class and others also moved. 

It was an extremely suprising and big change for a minor release (developer point of view). ;-)

**Comment by nknize** [comment]
@olfuerniss thank you for sharing your experiences. It's always helpful to get feedback from others in how they're using the artifacts. It sounds like you're developing your own plugin? Similar to your comments, the OpenSearch plugins in general deal with these kinds of issues every time-based minor release cycle for reasons outlined in #5910. As I mentioned, the refactoring effort is a move to fix this in the future to support an official Developer API and (since it's not there yet and requires all core maintainers pulling in the same direction) will require some rocky moments like this since the project attempts to minimize major releases.

Just to add a few things to open the aperture here:

> A breaking change info for developers would be important (+ necessary migration steps).

There is a [`>breaking`](https://github.com/opensearch-project/OpenSearch/pulls?q=is%3Apr+label%3A%3Ebreaking+is%3Aclosed) label that is supposed to be used for this purpose. Lot's of low hanging areas for improvement here. For example, labels (beyond just `>breaking`) could/should be better used and applied by maintainers (this is an area I've been trying hard to pull folks along), the changelog process is manual, non-standard, and clumsy (this was [discussed on a separate issue](https://github.com/opensearch-project/OpenSearch/issues/1868#issuecomment-1404252446), moved to a new [project wide proposal](https://github.com/opensearch-project/.github/issues/156), and months overdue on being revised. I think your experiences and wishes could help nudge that along to fix these communication struggles).


> It was an extremely suprising and big change for a minor release (developer point of view)

+1. Until the core API is defined (through JPMS, library exports, and other java mechanisms like `sealed` classes), plugin developers within and outside the opensearch project org are going to experience changes like this relatively often. It's a side effect of forking a code base that [did not establish a formal developer API](https://github.com/opensearch-project/OpenSearch/issues/5910) (for these exact painful reasons your describing). This is why the refactoring is happening to begin with but requires a delicate balance of how much deprecated code surface area and internal logic paths should be retained solely for backwards compatibility purposes vs moving to a well supported secure Generally Available API. On this project we talk about [SemVer in three areas](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#backwards-compatibility). Progress toward Developer API [enforcement](https://github.com/opensearch-project/OpenSearch/pull/9223) was only added as recent as two months ago (to assist with the API blast radius from PRs like this). But expectations need to be set. Annotations aren't yet fully applied, and where they are applied many will change over the coming months because of all the maintainers on core only one worked at elastic when the core plugin framework and xcontent classes were first designed and written so the annotation choice is only best effort.

**Pull request merged**
